sbin/ucored/ucored.8
.\"
.\" Copyright (c) 2025 Kyle Evans <kevans@FreeBSD.org>
.\"
.\" SPDX-License-Identifier: BSD-2-Clause
.\"
.Dd August 20, 2025
.Dt UCORED 8
.Os
.Sh NAME
.Nm ucored
.Nd process and filter userland coredump files
.Sh SYNOPSIS
.Nm
.Op Fl dLv
.Op Fl p Ar pidfile
.Sh DESCRIPTION
The
.Nm
daemon listens for coredump notifications from the kernel via
.Xr devd 8
and applies a policy to each generated coredump.
The default policy moves all coredumps to
.Pa /var/ucrash ,
though this is subject to change.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl d
Run in the foreground instead of detaching from the controlling terminal.
This is intended to be used in debugging
.Nm
itself.
All messages are redirected to either
.Va stdout
or
.Va stderr .
The
.Fl d
option does not imply any extra verbosity.
.It Fl L
Do not load
.Xr ucoredev 4
if it is not already loaded at startup.
.Nm
will otherwise try to load it and take over coredumps for the entire system.
This option implies a fallback to using
.Xr devd 8
notfications to hoover up coredumps if the module was not previously loaded.
.Pp
The
.Fl L
option has no effect if
.Xr ucoredev 4
does not exist.
.It Fl p Ar pidfile
Write the
.Nm
daemon pid to the specified
.Ar pidfile .
If the
.Ar pidfile
already exists and contains a valid pid, then
.Nm
will refuse to start.
.It Fl v
Run with verbosity.
The
.Fl v
option may be specified multiple times to increase verbosity.
At this time, specifying more verbosity than
.Fl vv
will have no effect.
.El
.Pp
Note that
.Nm
can run either standalone or initiated via
.Xr inetd 8
configured to handle its UNIX socket at
.Pa /var/run/ucored.sock .
.Pp
If
.Nm
is invoked via some socket-initiation launcher like
.Xr inetd 8 ,
then the
.Va kern.coredump_devctl
.Xr sysctl 8
will need to be manually enabled.
During standalone operation,
.Nm
will enable and disable it as needed.
.Ss Policies
The
.Nm
daemon offers some flexibility in the policies it will apply to a coredump.
The following policy actions are supported:
.Bl -tag -width indent
.It Dv discard
Discards the coredump completely.
.Nm
will promptly
.Xr unlink 2
the coredump if necessary, but cores produced by
.Xr ucoredev 4
do not exist on disk and do not need to be unlinked.
.Pp
This is a final rule, and must be the last rule specified in a chain.
.It Dv execute
Executes a
.Ar command .
The
.Ar command
must be an absolute path, as the working directory of
.Nm
is undefined.
The
.Ar command
may either be a string that will be word-split based on standard shell rules, or
an array of arguments.
.Pp
The process will be run with
.Pa /dev/null
hooked up to standard input, standard output, and standard error.
.Pp
The environment it will run in is a subset of the environment that
.Nm
runs in.
By default, only the
.Ev LANG
and
.Ev PATH
variables are inherited.
An
.Ar env
table may be supplied to augment the inherited environment.
Values of
.Ar env
may use shell-style
.Dq $VAR
substitution, as well as any of the substitutions described below in
.Sx Path Substitutions .
Shell-style variables will only be resolved from the running environment, and
will not be affected by other assignments specified in
.Ar env .
.It Dv ignore
Ignores the coredump.
For coredumps produced by
.Xr ucoredev 4 ,
this is the same as the
.Dv discard
policy.
.Pp
It does not make sense to combine an
.Dv ignore
rule with any other rule, and attempts to do so will raise an error.
.It Dv move
Moves the core file to some other
.Ar destination
on disk.
.Pp
If the specified
.Ar destination
does not have any path separators, then the operation is considered to be a
rename.
.Pp
If the
.Ar destination
is a relative path, then it is considered to be relative to the original
location of the coredump.
.Pp
If the
.Ar destination
does not end with a trailing slash, then the last component is assumed to be
a filename unless the specified path already exists and is a directory.
.Pp
See
.Sx Path Substitutions
below for more details on how
.Nm
may augment paths.
.It Dv pipe
Pipes the core file to another
.Ar command .
The
.Ar command
will be interpreted as it is specified for the
.Dv execute
action above, and the same constraints apply.
The core file will be opened, if necessary, and passed to the standard input of
the specified
.Ar command
instead of
.Pa /dev/null .
The notes for the
.Dv execute
action about the environment and
.Ar env
also apply to the
.Dv pipe
action.
.It Dv script
Specifies a script that will process the coredump.
The
.Ar file
specified should be a lua script that returns a function that will handle an
internal
.Dv ucore
object.
See
.Sx ucore Objects
below for more details on what
.Nm
provides in a ucore object.
.Pp
This is a final rule, and must be the last rule specified in a chain.
.El
.Ss Path Substitutions
The
.Nm
daemon will process paths that are supplied as a destination to
.Dv move
to and apply requested substitutions for data from the coredump notification.
These substitutions are applied to all parts of the path, not just the filename
component.
The following substitutions may be requested:
.Bl -tag -width indent
.It Dq \&%d
The domainname of the jail that the process belonged to, or the host if the
process was unjailed.
If unset, this will expand to the empty string.
.It Dq \&%f
The current path to the corefile.
.It Dq \&%h
The hostname of the jail that the process belonged to, or the host if the
process was unjailed.
If unset, this will expand to the empty string.
.It Dq \&%I
An index to uniquify a corefile.
Indices must be placed in the filename portion of a path.
It will not be replaced in any other position, and only the first occurrence
will be replaced.
.Pp
.Nm
will replace the
.Dq \&%I
token with an index from 0 up to the
.Va max_cores
specified on the action, or 5 if unspecified, until it finds a file that does
not exist.
If all of the allowed corefiles under this scheme exist, then the oldest one
will be replaced.
.Pp
Note that
.Nm
does not currently handle potential races well, so hopefully the process isn't
repeatedly crashing and every single dump needs to be captured.
.It Dq \&%j
The jid of the jail that the process belonged to.
This will be 0 for unjailed processes.
.It Dq \&%n
The name of the process.
.It Dq \&%P
The pid of the process' parent.
.It Dq \&%p
The pid of the process.
.It Dq \&%s
The signal that triggered the coredump.
.It Dq \&%u
The uid of the process credentials.
.El
.Ss ucore Objects
When information about a coredump has been received,
.Nm
constructs a
.Dq ucore
object with the relevant information.
The following methods are available on the ucore object:
.Bl -tag -width indent
.It Dv ucore:attributes()
Returns a table of attributes received.
The following attributes are recorded:
.Bl -tag -width indent
.It Dv compression
A string describing the compression of the core.
The following values are valid:
.Bl -tag -width indent
.It Dq
No compression.
This is
.Va kern.compress_user_cores Ns = Ns 0 .
.It Dq gzip
Compressed with gzip.
This is
.Va kern.compress_user_cores Ns = Ns 1 .
.It Dq zstd
Compressed with zstd.
This is
.Va kern.compress_user_cores Ns = Ns 2 .
.It Dq ???
Unknown compression state.
This shouldn't really happen in practice.
.El
.It Dv jid
The jid of the jail that the process belonged to.
This will be 0 for unjailed processes.
.It Dv gid
The gid of the process credentials.
.It Dv ppid
The pid of the process' parent.
.It Dv pid
The pid of the process.
.It Dv tainted
True if the process is running a setuid/setgid binary, or if the process had
changed credentials since
.Xr execve 2
time.
The
.Dv uid
and
.Dv gid
are overridden to root/wheel when a file is moved if the process has been
tainted, for security reasons.
.It Dv signal
The signal that triggered the coredump.
.It Dv uid
The uid of the process credentials.
.El
.It Dv ucore:comm()
The command that was running.
Note that the
.Dq comm
field in
.Fx
is limited to 19 bytes.
.It Dv ucore:jail()
The name of the jail the process was running in.
.It Dv ucore:filename()
The filename component of the core's path.
.It Dv ucore:move(path)
Move the core to the specified
.Ar path .
For file moves,
.Nm
will attempt to use
.Xr rename 2
first.
If the
.Xr rename 2
operation fails because it's a cross-filesystem rename, then
.Nm
will fallback to using
.Xr copy_file_range 2
to try and be more performant.
The owner, mode, and some file flags from the original coredump are preserved
in the move.
Specifically, the
.Dv nouchg
and
.Dv nodump
flags are preserved in case some other system policy would have set them.
.Pp
Note that
.Nm
may fail the move for security reasons in some circumstances.
If the coredump path provided by the kernel is now a symlink instead of a
regular file, then
.Nm
will refuse to move it.
If the destination exists and is not owned by the same user as the original
coredump, then
.Nm
will refuse to move it.
If the destination exists and is a symlink, then
.Nm
will also refuse to move it.
.It Dv ucore:path()
The full path to the coredump.
.El
.Ss Configuration
.Nm
is configured with UCL using
.Xr libucl 3 .
Its configuration, by default, is located at
.Pa /usr/local/etc/ucored.conf .
.Pp
The top-level object in the configuration is expected to have a
.Dq filters
key whose value is one or more filter objects.
Either a single object or an array of object(s) is accepted.
.Pp
Each filter object may have three keys:
.Bl -tag -width indent
.It Dv name
Optional.
A name to be used for reference in errors related to the filter.
If a name is not specified, then the index into hte filters array will be used
instead.
.It Dv match
Required.
An object whose keys describe fields to match on, and whose values contain
POSIX Extended Regular Expressions to match against.
All fields that are specified are tested, and a rule is only selected if all of
the specified expressions match.
The first rule to match will handle the corefile.
.Pp
The following fields are valid for matching against:
.Bl -tag -width indent
.It Dv comm
The command that was executed.
The comm field is the full path to the command when it's available.
.It Dv domainname
The domainname from the process' jail or the host, as applicable.
.It Dv hostname
The hostname from the process' jail or the host, as applicable.
.It Dv jail
The jail name from the process' jail if it was jailed, or the empty string if
it was not.
.It Dv path
The original path of the corefile.
For cores that were dumped through
.Xr ucoredev 4 ,
the path will be empty.
.El
.It Dv execute
Required.
One or more actions to execute against the core.
Either a single object or an array of object(s) is accepted.
.Pp
Each action object has a required key,
.Dv type ,
to name the action to execute.
The valid action types and the required arguments are described above in
.Sx Policies .
The named arguments are expected to be provided as a key in the same
object.
.El
.Sh FILES
.Pa /usr/local/etc/ucored.conf ,
.Pa /var/run/ucored.sock ,
.Pa /var/ucrash
.Sh EXIT STATUS
The
.Nm
utility exits 1 if an error occurs in startup.
.Sh SEE ALSO
.Xr chflags 1 ,
.Xr copy_file_range 2 ,
.Xr rename 2 ,
.Xr unlink 2 ,
.Xr pidfile 3 ,
.Xr signal 3 ,
.Xr ucoredev 4 ,
.Xr devd 8 ,
.Xr inetd 8 ,
.Xr sysctl 8


sbin/ucored/ucored.c
/*-
 * Copyright (c) 2025 Kyle Evans <kevans@FreeBSD.org>
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <sys/param.h>
#include <sys/event.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/sysctl.h>
#include <sys/un.h>
#include <sys/wait.h>

#include <assert.h>
#include <err.h>
#include <errno.h>
#include <signal.h>
#include <stdatomic.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include <libutil.h>

#include "ucored.h"

/* SOCK_CLOFORK wasn't introduced until 15.0. */
#ifndef SOCK_CLOFORK
#define SOCK_CLOFORK 0
#endif

/* sysctl to enable devctl notifications upon coredump. */
#define	DEVCTL_SYSCTL	"kern.coredump_devctl"

struct ucored_server {
	struct ucore_readable	serv_readable;
	int			serv_kq;
};

#define	UCORED_SERVER_OF(ur)	\
    __containerof(ur, struct ucored_server, serv_readable)

static int ucored_loop(int);
static bool ucored_accept(struct ucore_readable *, size_t, bool);

sig_atomic_t ucored_checkpwait;
sig_atomic_t ucored_terminate;

static void
handle_sigchld(int signo __unused)
{
	ucored_checkpwait = 1;
	atomic_signal_fence(memory_order_release);
}

static void
handle_termsig(int signo __unused)
{
	ucored_terminate = 1;
	atomic_signal_fence(memory_order_release);
}

static void
ucored_drain_zombies(void)
{
	pid_t wpid;
	int status;

	while ((wpid = waitpid(0, &status, WNOHANG)) != 0) {
		if (wpid == -1) {
			if (errno == EINTR)
				continue;
			if (errno == ECHILD)
				break;

			__assert_unreachable();
		}

		if (!WIFEXITED(status) || WEXITSTATUS(status) != 0) {
			libucore_log(LOG_NOTICE,
			    "Child %d terminated abnormally (status=%x)",
			    wpid, status);
			continue;
		}

		libucore_log(LOG_INFO, "Child %d terminated OK", wpid);
	}
}

static void
ucored_signal_setup(void)
{
	sigset_t set;
	struct sigaction sa = {
		/* No SA_RESTART */
		.sa_handler = handle_termsig,
	};

	(void)sigaction(SIGINT, &sa, NULL);
	(void)sigaction(SIGTERM, &sa, NULL);

	sa.sa_handler = handle_sigchld;
	(void)sigaction(SIGCHLD, &sa, NULL);

	sigemptyset(&set);
	sigaddset(&set, SIGINT);
	sigaddset(&set, SIGTERM);
	(void)sigprocmask(SIG_UNBLOCK, &set, NULL);
}

static int
ucored_sock(void)
{
	struct sockaddr_un sun;
	size_t ret;
	int sock;

	sock = socket(PF_UNIX, SOCK_STREAM | SOCK_CLOFORK, 0);
	if (sock == -1) {
		libucore_log(LOG_ERR, "socket: %m");
		return (-1);
	}

	if (fchmod(sock, 0600) == -1) {
		libucore_log(LOG_ERR, "fchmod: %m");
		close(sock);
		return (-1);
	}

	sun.sun_family = AF_UNIX;
	ret = strlcpy(&sun.sun_path[0], PATH_UCORED_SOCK, sizeof(sun.sun_path));
	assert(ret < sizeof(sun.sun_path));

	sun.sun_len = SUN_LEN(&sun);

	(void)unlink(PATH_UCORED_SOCK);
	if (bind(sock, (const struct sockaddr *)&sun, sizeof(sun)) == -1) {
		libucore_log(LOG_ERR, "bind: %m");
		close(sock);
		return (-1);
	}

	return (sock);
}

int
ucored_watch_socket(int kq, struct ucore_readable *ur)
{
	struct kevent ev;
	int fd = ur->r_fd;

	/*
	 * The continuous monitoring case is likely the listeing socket, since
	 * we don't really have a need to do that.  Clients will generally do
	 * one-shot for each segment of data they expect to receive.
	 */
	if (!ur->r_oneshot) {
		/* Listening socket */
		EV_SET(&ev, fd, EVFILT_READ, EV_ADD, 0, 0, ur);
	} else if (ur->r_lowat != NULL) {
		EV_SET(&ev, fd, EVFILT_READ, EV_ADD | EV_ONESHOT,
		    NOTE_LOWAT, (*ur->r_lowat)(ur), ur);
	} else {
		EV_SET(&ev, fd, EVFILT_READ, EV_ADD | EV_ONESHOT, 0, 0, ur);
	}

	if (kevent(kq, &ev, 1, NULL, 0, NULL) == -1) {
		libucore_log(LOG_ERR, "kevent: %m");
		return (-1);
	}

	return (0);
}

/*
 * Sets the sysctl to newst and returns the previous state.  We'll be a good
 * citizen and disable it on our way out.
 */
static int
devctl_fiddle(int newst)
{
	int oldval;
	size_t oldlen;

	oldlen = sizeof(oldval);
	if (sysctlbyname(DEVCTL_SYSCTL, &oldval, &oldlen, &newst,
	    sizeof(newst)) == -1) {
		libucore_log(LOG_ERR, "sysctlbynme: %m");
		return (-1);
	}

	return (oldval);
}

static void
usage(void)
{

	fprintf(stderr, "usage: %s [-dv] [-p pidfile]\n", getprogname());
	exit(1);
}

static bool
ucored_socket_check(int fd)
{
	struct stat sb;

	/*
	 * If we can't tell that it's a socket, then we err on the side of
	 * caution and assume it's not.  Otherwise, we'll operate on it.
	 */
	if (fstat(fd, &sb) == -1)
		return (false);
	return (S_ISSOCK(sb.st_mode));
}

int
main(int argc, char *argv[])
{
	const char *pidfile = NULL;
	struct pidfh *pidfh = NULL;
	struct ucore_dev *udev = NULL;
	struct ucored_server *userv = NULL;
	int ch, error = 1, devctl_prev, kq = -1, sock = -1, verbose = 0;
	bool debug = false, loadkmod = true, socket_initiated;

	while ((ch = getopt(argc, argv, "dLp:v")) != -1) {
		switch (ch) {
		case 'd':
			debug = true;
			break;
		case 'L':
			loadkmod = false;
			break;
		case 'p':
			pidfile = optarg;
			break;
		case 'v':
			verbose++;
			break;
		default:
			usage();
			break;
		}
	}

	argc -= optind;
	argv += optind;

	libucore_set_debug(debug);
	libucore_set_verbose(verbose);

	if (argc != 0)
		usage();

	/*
	 * Check if we're socket initiated, e.g., inetd-style.
	 */
	socket_initiated = ucored_socket_check(STDIN_FILENO);
	if (socket_initiated)
		sock = STDIN_FILENO;

	if (pidfile != NULL &&
	    (pidfh = pidfile_open(pidfile, 0644, NULL)) == NULL) {
		if (errno == EEXIST)
			errx(1, "ucored is already running");

		warn("pidfile_open: %m");
	}

	if (!ucored_lua_init()) {
		pidfile_remove(pidfh);
		return (1);
	}

	if (!debug && !socket_initiated && daemon(0, 0) == -1) {
		fprintf(stderr, "daemon: %s", strerror(errno));
		pidfile_remove(pidfh);
		return (1);
	}

	pidfile_write(pidfh);

	kq = kqueue();
	if (kq == -1) {
		libucore_log(LOG_ERR, "kqueue: %m");
		goto done_nosock;
	}

	if (!socket_initiated) {
		struct ucore_readable *ur;

		/*
		 * XXX Should the socket and device be mutually exclusive?
		 */
		if (libucore_dev_available(loadkmod)) {
			udev = libucore_dev_open(ucored_lua_handle);
			if (udev == NULL) {
				libucore_log(LOG_ERR, "failed to open /dev/ucore");
				goto done;
			}

			ur = libucore_dev_readable(udev);
			if (ucored_watch_socket(kq, ur) == -1)
				goto done;

			libucore_log(LOG_INFO, "ucored polling /dev/ucore");
		} else {
			sock = ucored_sock();
			if (sock == -1)
				goto done_nosock;

			if (listen(sock, 10) == -1) {
				libucore_log(LOG_ERR, "listen: %m");
				goto done;
			}

			userv = calloc(1, sizeof(*userv));
			if (userv == NULL) {
				libucore_log(LOG_ERR, "calloc: %m");
				goto done;
			}

			userv->serv_kq = kq;

			ur = &userv->serv_readable;
			ur->r_read = ucored_accept;
			ur->r_fd = sock;

			if (ucored_watch_socket(kq, ur) == -1)
				goto done;
		}
	} else if (ucored_client_alloc(kq, sock) == NULL) {
		goto done;
	}

	/*
	 * We'll setup SIGINT/SIGTERM to interrupt us, after which we'll briefly
	 * evacuate the daemon and clean up our current state.  Any cores that
	 * are in the process of being handled will be finished, but if we were
	 * in the middle of receiving core metadata we'll drop that one on the
	 * floor.
	 */
	ucored_signal_setup();

	if (!socket_initiated && sock >= 0)
		devctl_prev = devctl_fiddle(1);

	error = ucored_loop(kq);

	/*
	 * Cleanup time; make sure we don't leave our socket laying around, and
	 * we'll also want to restore the previous system state for whether or
	 * not the kernel generates devctl notifications for coredumps.
	 */
	if (!socket_initiated && sock >= 0 && !devctl_prev)
		(void)devctl_fiddle(0);

	ucored_client_close_all();

done:
	if (!socket_initiated && sock >= 0)
		(void)unlink(PATH_UCORED_SOCK);
done_nosock:
	pidfile_remove(pidfh);

	if (kq >= 0)
		close(kq);
	libucore_dev_close(udev);
	free(userv);
	if (sock >= 0)
		close(sock);
	return (error);
}

void
ucored_now(struct timespec *tsp)
{
	(void)clock_gettime(CLOCK_UPTIME, tsp);
}

static bool
ucored_accept(struct ucore_readable *ur, size_t backlog, bool eof __unused)
{
	struct ucored_server *userv = UCORED_SERVER_OF(ur);
	int clsock;
	int fd = ur->r_fd, kq = userv->serv_kq;

	/*
	 * All errors here are considered non-fatal.  We'll return and move on
	 * in case it's a transient condition; perhaps some other event we will
	 * be executing on would relieve it.  In the worst case scenario, all
	 * we have are new connections and we'll keep returning here just to
	 * shut them down.
	 */
	for (size_t i = 0; i < backlog; i++) {
		clsock = accept4(fd, NULL, NULL, SOCK_CLOFORK);
		if (clsock == -1) {
			libucore_log(LOG_ERR, "accept: %m");
			return (true);
		}

		if (ucored_client_alloc(kq, clsock) == NULL)
			return (true);
	}

	return (true);
}

static void
ucored_timer(void)
{
	size_t purged;

	purged = ucored_client_purge_inactive();
	if (purged > 0)
		libucore_log(LOG_NOTICE, "purged %zu inactive clients", purged);
}

static bool
ucored_set_timer(int kq, bool state)
{
	struct kevent kev;
	int flag;

	flag = EV_ADD | (state ? EV_ENABLE : EV_DISABLE);
	EV_SET(&kev, 0, EVFILT_TIMER, flag, NOTE_SECONDS, UCORED_TIMEOUT + 1,
	    NULL);
	if (kevent(kq, &kev, 1, NULL, 0, NULL) == -1) {
		libucore_log(LOG_ERR, "%s timer kevent: %m",
		    state ? "enable" : "disable");
		return (false);
	}

	return (true);
}

static int
ucored_loop(int kq)
{
	struct kevent kev[4];
	int ret;

	assert(kq >= 0);

	for (;;) {
		size_t current_clients;
		bool timer_fired;

		atomic_signal_fence(memory_order_acquire);

		if (ucored_checkpwait) {
			ucored_drain_zombies();
			ucored_checkpwait = 0;
		}

		if (ucored_terminate)
			break;

		ret = kevent(kq, NULL, 0, &kev[0], nitems(kev), NULL);
		if (ret == -1) {
			if (errno == EINTR)
				continue;

			libucore_log(LOG_ERR, "kevent: %m");
			return (1);
		}

		current_clients = ucored_clients;
		timer_fired = false;
		for (int idx = 0; idx < ret; idx++) {
			const struct kevent *evt = &kev[idx];
			struct ucore_readable *ur;
			int fd;
			bool rearm, oneshot;

			fd = evt->ident;
			ur = evt->udata;

			/*
			 * Process any other events before we process the timer, just in
			 * case one of them is a client trying to send us information.
			 */
			if (evt->filter == EVFILT_TIMER) {
				timer_fired = true;
				continue;
			}

			assert(ur != NULL);

			libucore_log(LOG_DEBUG, "Fetching data from fd %d",
			    fd);

			oneshot = ur->r_oneshot;
			rearm = (*ur->r_read)(ur, evt->data,
			    (evt->flags & EV_EOF) != 0);

			/*
			 * Client may not be valid anymore, unless we were
			 * asked to re-arm.  If we were asked to terminate, just
			 * bail out anyways.
			 */
			if (ucored_terminate)
				goto out;

			/*
			 * The readable needs to coordinate its own removal if
			 * it wants to go away; we don't manage the lifetime of
			 * ucore_readable objects here.  We stashed r_oneshot
			 * away just in case the object was freed during the
			 * r_read() callback, but if it requested re-arm then
			 * the readable should still be alive.
			 */
			if (!oneshot) {
				assert(rearm);
				continue;
			}

			if (rearm)
				ucored_watch_socket(kq, ur);
		}

		if (timer_fired)
			ucored_timer();

		/* On a state transition, arm or disarm our timer. */
		if ((current_clients != 0) != (ucored_clients != 0)) {
			bool state = ucored_clients != 0;

			libucore_log(LOG_INFO,
			    "Clients went from %zu -> %zu, %sarming the alarm",
			    current_clients, ucored_clients,
			    state ? "" : "dis");
			ucored_set_timer(kq, state);
		}
	}

out:
	return (0);
}
